Docker support for all three interfaces(jax,tensorflow,pyorch) including CPU+GPU builds

arshpreetsingh opened the pull request:
**Context:**

Add Docker support to build/setup Pennylane in one Command. 

**Description of the Change:**

user would be able to build Pennylane and run Tests for specific Python and interface version. 

**Benefits:**
single step build for multiple flavours for Pytorch. 

**Possible Drawbacks:**
Need to test GPU builds on REAL-GPU devices.

**Related GitHub Issues:**
Closes https://github.com/PennyLaneAI/pennylane/issues/1372

co9olguy commented:
Thanks for this PR @arshpreetsingh, we'll have a member of the team take a look and provide an initial review :muscle:

gtr8 commented on docker/gpu-cuda/jax.dockerfile:
```diff
@@ -0,0 +1,46 @@
+# Setup Miniconda
+FROM dev-base as conda
+ARG PYTHON_VERSION=3.7
+RUN curl -fsSL -v -o ~/miniconda.sh -O  https://repo.anaconda.com/miniconda/Miniconda3-py39_4.9.2-Linux-x86_64.sh  && \
+    chmod +x ~/miniconda.sh && \
```
@arshpreetsingh Thanks for the PR! Should we consider adding checksum verification to the miniconda installation script? 🤔 

https://docs.conda.io/en/latest/miniconda.html#linux-installers
https://conda.io/projects/conda/en/latest/user-guide/install/download.html#cryptographic-hash-verification

arshpreetsingh replied:
@gtr8 Yes, That option could be added. once it will be finalised either have to keep conda or not(licensing discussion) we can update accordingly.

mlxd commented on docker/jax.dockerfile:
```diff
@@ -0,0 +1,45 @@
+# Setup Miniconda
+FROM dev-base as conda
+ARG PYTHON_VERSION=3.7
+RUN curl -fsSL -v -o ~/miniconda.sh -O  https://repo.anaconda.com/miniconda/Miniconda3-py39_4.9.2-Linux-x86_64.sh  && \
```
```suggestion
RUN curl -fsSL -v -o ~/miniconda.sh -O  https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh  && \
```
It may be better to run with the latest miniconda version rather than a given version string.

arshpreetsingh replied:
Just wanted to know if it's good idea to go with Conda, conda builds takes lot of space like 1.6 GB but on the other hand builds using default Python is around 600MB which is huge space-save. I have added new Dockerfiles with some more robust method than last one and pushed all builds using Conda into separate folder.     

Will it be good idea to keep Conda builds separate?

mlxd replied:
Good question, and great job on the changes! I think for now, keeping them separate is a good idea. Given the large difference in size, I think I am in favour of the smaller images, though we can keep both for the moment and see are there situations where Conda makes more sense.

mlxd reviewed (changes requested):
Hi @arshpreetsingh thanks for the work on this! It will be great to get fully-featured Docker support in here. I do have some comments and requests:
- The use of Conda does simplify the install for many packages. Though, I do wonder if this is something we need to consider from a licensing perspective. It may be better overall to favour a pip-installed docker image environment. We will discuss this internally and get back to you.
- On the above, if conda is chosen, it can be quite heavyweight, and vastly contribute to image size (the base pennylane image for example is 1.4GB). It would be good to remove all non-essential packages and binaries, and to prefer smaller versions of libraries (for a few examples: remove the miniconda installer and all intermediate and non-required conda-packages, prefer OpenBLAS over MKL for BLAS library, avoid updating existing packages in conda). Ideally, a base image of a few hundred MB at most would be best.
- Is there any reason why Ubuntu 18:04 is chosen rather than the latest LTS 20:04?
- Many of the image files contain a lot of overlapping commands. Even though I know the docker env will cache these commands, it would be great to have a base image file that is used by all of these, as it will cut down on repeated edits and help with maintenance in the future.
- The use of a makefile does simplify use and builds, but this will not be capable of running on a Windows machine (assuming the user does not have WSL). I think the use of the cross-platform compatible [docker-compose](https://docs.docker.com/compose/) will help with this, as all commands can be run on any machine that supports docker. This also opens up the possibility of much richer configuration and usage options.
- On the above, right now all images copy the pennylane directory onto the image. While this is great for getting up and running, any changes on the image will vanish when the user closes the running container. I think volume mounting the pennylane directory onto the image would be a better approach here, as all changes persist on the user's filesystem. This can be easily accomplished with the docker-compose file.

I haven't had time to test the GPU images, but will try to get to this soon on a CUDA machine.

arshpreetsingh commented on docker.makefile:
```diff
@@ -0,0 +1,57 @@
+#Dockerfile vars
```
@mlxd  agreed with docker-compose, docker-compose is more robust, But I believe Makefile implementation should also be there in case someone wants to do things  using Kubernetes or  just doing build-testing for the local ENV only.

mlxd replied:
Thanks @arshpreetsingh While I agree that the makefile implementation is definitely useful, the issue is more for cross-platform use as they can only be used on Linux/Mac machines --- I would also expect deploying to a Kubernetes cluster to be more in-line with the docker-dompose syntax, given that both are configured in YAML, though I think Kubernetes support can come a little after Docker support in a future PR :slightly_smiling_face:  

If we intend to add support for Windows, we should ensure the users have a way to access this functionality on a Windows machine, or really any machine on which Docker supports.

Also, I think it would be good even with the makefile/docker-compose.yml, to list the commands required to access the PennyLane image as a runnable environment using docker: something like `docker run -it pennylane:latest /bin/bash` for those not familiar with running docker containers.

This gives us a few potentially different options to explore:
- Keep the makefile, and add a comment to the README to use the makefile with WSL under Windows if Docker support is required.
- Keep the makefile and add an equivalent Powershell file for building the images on Windows, with similar functionality as the makefile.
- Replace the makefile with a cross-platform docker-compose.yml file (also allows easy building, running and mounting of the cloned PL repository into the image).
- Explicitly list the build & run commands on the README for each given image the user wishes to use.

For now, I think the quickest change may be to add the comments to the README for using WSL under Windows.

arshpreetsingh replied:
@mlxd  So should we talk about docker-compose implementation  in this PR or in other one?

arshpreetsingh commented on docker/gpu-cuda/tensorflow.dockerfile:
```diff
@@ -0,0 +1,49 @@
+
+ARG BASE_IMAGE=nvidia/cuda:11.1-base
+
+# Setup develop base image packages(build-essentials etc)
+FROM ${BASE_IMAGE} as dev-base
+
+RUN apt-get update && apt-get install -y --no-install-recommends
+RUN DEBIAN_FRONTEND="noninteractive" apt-get install tzdata
+RUN apt-get install -y build-essential \
+        tzdata \
+        ca-certificates \
+        ccache \
+        cmake \
+        curl \
+        git \
+        libjpeg-dev \
+        libpng-dev && \
+    rm -rf /var/lib/apt/lists/*
+RUN /usr/sbin/update-ccache-symlinks
+RUN mkdir /opt/ccache && ccache --set-config=cache_dir=/opt/ccache
+ENV PATH /opt/conda/bin:$PATH
+
+# Setup Miniconda
+FROM dev-base as conda
+ARG PYTHON_VERSION=3.7
+RUN curl -fsSL -v -o ~/miniconda.sh -O  https://repo.anaconda.com/miniconda/Miniconda3-py39_4.9.2-Linux-x86_64.sh  && \
+    chmod +x ~/miniconda.sh && \
+    ~/miniconda.sh -b -p /opt/conda && \
+    rm ~/miniconda.sh && \
+    /opt/conda/bin/conda clean -ya
+
```
Yes that repetition has been removed in recent PUSH.

mlxd replied:
Great work on this; I really like the simplified Dockerfiles --- I think we should opt for these in-place of the Conda files.

arshpreetsingh commented on docker.makefile:
```diff
@@ -0,0 +1,57 @@
+#Dockerfile vars
+#
+# ARG BASE_IMAGE=18.04
+# ARG PYTHON_VERSION=3.7
+# ARG TENSORFLOW_VERSION=2.5
+# ARG PyTorch_VERSION=1.8.1
+# #ARG Jax_VERSION=
```
Removed those comments, Also  added ```make -f docker.makefile build-all``` to build all 3 interfaces(torch,tensorflow,jax) together using single command.

mlxd replied:
Great, thanks!

arshpreetsingh commented:
> Hi @arshpreetsingh thanks for the work on this! It will be great to get fully-featured Docker support in here. I do have some comments and requests:
> 
> * The use of Conda does simplify the install for many packages. Though, I do wonder if this is something we need to consider from a licensing perspective. It may be better overall to favour a pip-installed docker image environment. We will discuss this internally and get back to you.
> * On the above, if conda is chosen, it can be quite heavyweight, and vastly contribute to image size (the base pennylane image for example is 1.4GB). It would be good to remove all non-essential packages and binaries, and to prefer smaller versions of libraries (for a few examples: remove the miniconda installer and all intermediate and non-required conda-packages, prefer OpenBLAS over MKL for BLAS library, avoid updating existing packages in conda). Ideally, a base image of a few hundred MB at most would be best.
> * Is there any reason why Ubuntu 18:04 is chosen rather than the latest LTS 20:04?
> * Many of the image files contain a lot of overlapping commands. Even though I know the docker env will cache these commands, it would be great to have a base image file that is used by all of these, as it will cut down on repeated edits and help with maintenance in the future.
> * The use of a makefile does simplify use and builds, but this will not be capable of running on a Windows machine (assuming the user does not have WSL). ~I think the use of the cross-platform compatible [docker-compose](https://docs.docker.com/compose/) may help with this, as all commands can be run on any machine that supports docker. This also opens up the possibility of much richer configuration and usage options.~ Maybe an equivalent Powershell file would help here, or simply mentioning the requirement for WSL in the README.
> * On the above, right now all images copy the pennylane directory onto the image. While this is great for getting up and running, any changes on the image will vanish when the user closes the running container. I think volume mounting the pennylane directory onto the image would be a better approach here, as all changes persist on the user's filesystem. ~This can be easily accomplished with the docker-compose file.~ This may be worth updating the readme with an example of how to achieve this, and removing the copy operations from the build.
> 
> I haven't had time to test the GPU images, but will try to get to this soon on a CUDA machine.
> 
> Let me know your thoughts on the above, and we can see what makes the most sense.

 Hi @mlxd 
1. Now building images without conda, all conda files kept at /docker/conda* to update if required, otherwise defaut Python build is around ~600 MB which I have updated.  
2. Now using ubuntu 20:04 
3. Overlapping commands issue is also resolved.
4. now every interface build will  use pennylane:latest as base image, i will help to save space in user's system. 
5. Yes, Agree with docker-compose, Will look for implementation methods.

mlxd commented on docker.makefile:
```diff
@@ -0,0 +1,53 @@
+			 	@docker build -t pennylane-torch-gpu -f docker/gpu-cuda/tensorflow.dockerfile .
+
+build-all:
+			 	@docker build -t pennylane -f docker/pennylane.dockerfile .
+				&& @docker build -t pennylane -f docker/tensorflow.dockerfile .
+				&& @docker build -t pennylane -f docker/pytorch.dockerfile .
+				&&  @docker build -t pennylane -f docker/jax.dockerfile .
```
The `build-all` command fails after building the pennylane base image. The above changes should fix that.

mlxd replied:
Also, the tagging option above overwrites the original base image on each subsequent build.

I think it best to choose a naming schema such as: 
- Base -> `pennylane/base`
- TF -> `pennylane/tensorflow`
- Jax -> `pennylane/jax`
- Torch -> `pennylane/torch`

mlxd reviewed (changes requested):
Great job on the changes! A few more small comments.

josh146 commented on README.md:
```diff
@@ -101,6 +101,41 @@ scripts.
+
+Build using CPU as well GPU (Nvidia CUDA 11.1+) is also supported.
+
+To build a basic PennyLane image without any additional interfaces (Torch, TensorFlow or Jax) run following:
```
```suggestion
To build a basic PennyLane image without any additional interfaces (Torch, TensorFlow, or Jax) run following:
```

mlxd commented on docker/qchem.dockerfile:
```diff
@@ -0,0 +1,52 @@
+WORKDIR /opt/pennylane/qchem
+COPY  . .
+RUN git submodule update --init --recursive
+RUN  pip install wheel && pip install psi4 psi4-rt python=python3 -c psi4 && pip install -r requirements.txt \
```
This command will fail, as Psi4 isn't available as a pip installable package. 
I'm currently investigating whether we can just use pyscf instead of psi4 here, as psi4 is conda-only, or build-from-source.

If PySCF isn't a good replacement, I'll help with setting up a build script for psi4, or we can use conda for this one image.

arshpreetsingh replied:
I tried with "pip install openfermionpsi4" but build is still failing.

mlxd replied:
Yes, openfermionpsi4 expects Psi4 to be installed separately, which is unfortunate, as this causes some challenges (as you have just seen :slightly_smiling_face: ). Psi4 also brings a hard requirement of Python3.7 with it. I'm currently chatting with members internally about the best approach here, and will update you as soon as we have a way forward.

mlxd replied:
Thanks for the patience on this. So a quick follow-up: it seems that we can remove psi4 from the qchem image as it is an optional package (as per https://github.com/PennyLaneAI/pennylane/tree/master/qchem), and opt instead for PySCF (which can be pip installed), and `openfermion-pyscf` should allow the same functionalities as Psi4 and openfermion-psi4. Rather than installing directly from the qchem `requirements.txt` I think it would be better to be explicit about which packages to install: namely, install all packages but psi4 and openfermionpsi4. This should allow the image to be built.

arshpreetsingh replied:
Yes, Tried that but still having issues while running "make test". it seems issue is with ENV path, will dig it tonight.

mlxd commented on docker/tensorflow.dockerfile:
```diff
@@ -0,0 +1,6 @@
+From pennylane/base:latest
+
+# Update and install TensorFlow
+RUN apt-get update && pip3 install tensorflow==2.5.0
```
Running the tests after installing these packages would be a good idea for all images.

mlxd commented on docker/torch.dockerfile:
```diff
@@ -0,0 +1,6 @@
+From pennylane/base:latest
+
+# Update and install Torch
+RUN apt-get update && pip3 install torch==1.8.1
```
As above.

mlxd commented on docker/tensorflow.dockerfile:
```diff
@@ -0,0 +1,6 @@
+# Update and install TensorFlow
+RUN apt-get update && pip3 install tensorflow==2.5.0
+# Image completed, Exit Now.
+CMD echo "Docker image Builing process is Successful"
```
```suggestion
CMD echo "Successfully built Docker image."
```

mlxd commented on docker/qchem.dockerfile:
```diff
@@ -0,0 +1,52 @@
+ENV PATH="/opt/venv/bin:$PATH"
+RUN apt-get update && apt-get install -y --no-install-recommends python3 python3-pip python3-venv
+# Image completed, Exit Now.
+CMD echo "Docker image Builing process is Successful"
```
```suggestion
CMD echo "Successfully built Docker image."
```

mlxd commented on docker/gpu-cuda/tensorflow.dockerfile:
```diff
@@ -0,0 +1,6 @@
+# Update and install TensorFlow
+RUN apt-get update && pip3 install tensorflow==2.5.0
+# Image completed, Exit Now.
+CMD echo "Docker image Builing process is Successful"
```
```suggestion
CMD echo "Successfully built Docker image"
```

mlxd commented on docker/gpu-cuda/jax.dockerfile:
```diff
@@ -0,0 +1,6 @@
+# Update and install Jax
+RUN apt-get update && pip3 install jax jaxlib==0.1.67+cuda111 -f https://storage.googleapis.com/jax-releases/jax_releases.html
+# Image completed, Exit Now.
+CMD echo "Docker image Builing process is Successful"
```
```suggestion
CMD echo "Successfully built Docker image"
```

mlxd commented on docker/gpu-cuda/tensorflow.dockerfile:
```diff
@@ -0,0 +1,6 @@
+From pennylane/cuda/base:latest
+
+# Update and install TensorFlow
+RUN apt-get update && pip3 install tensorflow==2.5.0
```
Tests must be rerun here.

mlxd commented on docker/gpu-cuda/jax.dockerfile:
```diff
@@ -0,0 +1,6 @@
+From pennylane/cuda/base:latest
+
+# Update and install Jax
+RUN apt-get update && pip3 install jax jaxlib==0.1.67+cuda111 -f https://storage.googleapis.com/jax-releases/jax_releases.html
```
Tests must be rerun.

Also:
```
jaxlib==0.1.67+cuda111
```
I think the above package version should be specified somewhere that we have ease-of-access to update. Is there a way to ensure we pull in a jaxlib with CUDA support that matches the jax version, without explicitly listing a version number? If not, we should also pin jax to the same version.

arshpreetsingh replied:
Sure, I will verify that part. as much I understand JAX has different versions fro CPU and  GPU or may be it's true for JAXLIB only. Will investigate this one.

mlxd commented on docker/gpu-cuda/cuda-base.dockerfile:
```diff
@@ -0,0 +1,41 @@
+ENV PATH="/opt/venv/bin:$PATH"
+RUN apt-get update && apt-get install -y --no-install-recommends python3 python3-pip python3-venv
+# Image completed, Exit Now.
+CMD echo "Docker image Builing process is Successful"
```
See comments in other images.

mlxd reviewed (changes requested):
Great job @arshpreetsingh. It looks like we are nearly there. I do have some comments and suggestions, with a quick summary here:

- I think we can remove the conda directory and use the pip images only (with maybe 1 exception)
- Since Psi4 is a conda-only package (not available via pip), this may be the only place we need a conda environment. However, we may be able to use another package for this instead -- I am chatting with the QChem team currently to see what works here.
- I think it would be best to rerun all tests after installing each backend interface. This will ensure we catch any issues that may have come from installing the new package.
- Testing of the CUDA build will also be important, as we need to ensure the packages are running on the GPUs. We may need to figure out a good way to test these before accepting the GPU builds.
- Otherwise, minor formatting and wording changes

arshpreetsingh commented on docker.makefile:
```diff
@@ -0,0 +1,63 @@
+	@docker build -t pennylane/cuda/base -f docker/gpu-cuda/cuda-base.dockerfile . \
+	&& docker build -t pennylane/tensorflow/gpu -f docker/gpu-cuda/tensorflow.dockerfile . \
+	&& docker build -t pennylane/torch/gpu -f docker/gpu-cuda/torch.dockerfile . \
+	&& docker build -t pennylane/jax/gpu -f docker/gpu-cuda/jax.dockerfile .
```
Resolved.

arshpreetsingh commented on docker/torch.dockerfile:
```diff
@@ -0,0 +1,6 @@
+# Update and install Torch
+RUN apt-get update && pip3 install torch==1.8.1
+# Image completed, Exit Now.
+CMD echo "Docker image Builing process is Successful"
```
Resolved

mlxd reviewed (changes requested):
Thanks again for the changes. A few follow-up comments:

arshpreetsingh commented on README.md:
```diff
@@ -101,6 +101,42 @@ scripts.
+
+
+Docker support exists for building using CPU and GPU (Nvidia CUDA 11.1+) images.
+
```
Mentioned the WSL thing for Windows. Please let me know if any changes required.

mlxd replied:
That change looks perfect, thanks!

mlxd reviewed (commented):
Just an update: the tests for the CPU images are all passing except for one. I'm currently relaying the message to the team to see about a resolution. 

In the mean-time, I'd recommend merging the recent release master into here, running the images locally, and posting any logging output you get for failures here, as this may help debug the issue. Thanks!

arshpreetsingh commented:
> Just an update: the tests for the CPU images are passing except for one. I'm currently relaying the message to the team to see >about a resolution.

Hi @mlxd  Thanks for the comment!
Yes only Non-Conda Qchem is failing for me, I also tried multiple ways to fix Qchem installation using default pip methods but it seems it works well only under Conda-ENV. 

Should we revert back to Conda only for Qchem  and try to find ways to reduce image size with Conda-implementation?      
 
> In the mean-time, I'd recommend merging the recent release master into here, running the images locally, and posting any 
> logging output you get for failures here, as this may help debug the issue. Thanks!

Sure I could continue that activity.

arshpreetsingh commented:
> Great job @arshpreetsingh. It looks like we are nearly there. I do have some comments and suggestions, with a quick summary here:
> 
> * I think we can remove the conda directory and use the pip images only (with maybe 1 exception)
> * Since Psi4 is a conda-only package (not available via pip), this may be the only place we need a conda environment. However, we may be able to use another package for this instead -- I am chatting with the QChem team currently to see what works here.
> * I think it would be best to rerun all tests after installing each backend interface. This will ensure we catch any issues that may have come from installing the new package.
> * Testing of the CUDA build will also be important, as we need to ensure the packages are running on the GPUs. We may need to figure out a good way to test these before accepting the GPU builds.
> * Otherwise, minor formatting and wording changes



> Just an update: the tests for the CPU images are passing except for one. I'm currently relaying the message to the team to see about a resolution.
> 
> In the mean-time, I'd recommend merging the recent release master into here, running the images locally, and posting any logging output you get for failures here, as this may help debug the issue. Thanks!

@mlxd  any update on from Qchem team related to default PIP instead of Conda, Or should we move with conda-only for Qchem?

mlxd commented:
Hi @arshpreetsingh. The team had informed me that the pip approach should work, but there is still some need to investigate the cause of the failures. Though, as a quick aside, the line `COPY  . .` in the qchem file is not necessary, as it copies and overwrites essential files that were previously copied as part of the PennyLane base install --- this will then corrupt the install. I would recommend removing this line, running the testsuite on your local machine, then report the observed failing tests in the chat here (logs and everything). It should be possible to tag the respective people in to assist. Thanks!

arshpreetsingh commented:
> Hi @arshpreetsingh. The team had informed me that the pip approach should work, but there is still some need to investigate the cause of the failures. Though, as a quick aside, the line `COPY . .` in the qchem file is not necessary, as it copies and overwrites essential files that were previously copied as part of the PennyLane base install --- this will then corrupt the install. I would recommend removing this line, running the testsuite on your local machine, then report the observed failing tests in the chat here (logs and everything). It should be possible to tag the respective people in to assist. Thanks!

Hi @mlxd , This is resolved! thanks for Amazing suggestion. Qchem Build is resolved now!

arshpreetsingh commented:
@josh146  @mlxd  Closed by Mistake. Reopening!

mlxd commented on README.md:
````diff
@@ -101,6 +101,45 @@ scripts.
+To build a basic PennyLane image without any additional interfaces (Torch, TensorFlow, or Jax) run following:
+
+```
+make -f docker/docker.makefile build-base
````
This fails to run with error: `docker/docker.makefile:19: *** missing separator.  Stop.` and also prevents all the other images from currently being built and tested.

mlxd commented on docker/docker.makefile:
```diff
@@ -0,0 +1,33 @@
+		@echo "build-interface-gpu"
+
+build-base:
+    @docker build -t pennylane/base -f docker/pennylane.dockerfile .
```
Something is wrong here as the build is failing on my machine at this line. Also, please ensure the indentation level matches throughout :slightly_smiling_face:

mlxd commented on docker/docker.makefile:
```diff
@@ -0,0 +1,33 @@
+		&& docker build -t pennylane/qchem -f docker/qchem.dockerfile .
+
+build-interface:
+	  @docker build -t pennylane/base -f docker/pennylane.dockerfile . \
```
Indentation level differs here from above

mlxd commented on docker/docker.makefile:
```diff
@@ -0,0 +1,33 @@
+		--build-arg INTERFACE_NAME=$(interface-name) .
+
+build-interface-gpu:
+    @docker build -t pennylane/base -f docker/pennylane.dockerfile . \
```
Indentation also different here.

mlxd commented on docker/gpu-cuda/build_interface_gpu.dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+# Build TensorFlow interface
+RUN if [ "$INTERFACE_NAME" = "tensorflow" ] ; then \
+      apt-get update && apt-get -y install --no-install-recommends make \
+      && rm -rf /var/lib/apt/lists/* && pip3 install tensorflow==2.5.0;fi
```
Indentation here different from above.

mlxd commented on docker/gpu-cuda/build_interface_gpu.dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+# Build Torch interface
+RUN if [ "$INTERFACE_NAME" = "torch" ] ; then \
+      apt-get update && apt-get -y install --no-install-recommends make \
+      && rm -rf /var/lib/apt/lists/* && pip3 install torch==1.8.1;fi
```
Indentation here also different.

mlxd commented on docker/qchem.dockerfile:
```diff
@@ -0,0 +1,16 @@
+WORKDIR /opt/pennylane/qchem
+RUN git submodule update --init --recursive
+RUN pip install wheel && pip install openfermionpyscf && pip install -r requirements.txt \
+        && python3 setup.py install \
+        && make test
```
Indentation here differs from above.

mlxd reviewed (changes requested):
Hi @arshpreetsingh. Thanks again for your contributions. I have some notes:

I am unable to test any of the images currently as the base build is failing. I think the documentation needs to be updated to match the recent changes, following this fix so that we can build and test each image. I think it would be great if all the images can build and run the tests before the next review.

Also, the dockerfiles themselves differ a lot in indentation level. I think this should be standardised across the PR and following the examples as given in the Docker [best-practices guide](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/).

antalszava commented:
Hi @arshpreetsingh, just double-checking: should we go for another review round here or are some further changes planned before that?

arshpreetsingh commented:
> Hi @arshpreetsingh, just double-checking: should we go for another review round here or are some further changes planned before that?

Hi @antalszava  @mlxd  
All changes done now. 
1. Now using single dockerfile to build all three interfaces(torch, tensorflow, jax)
2. Updated Makefile and indentation issues for Docker.
3. Added Plugins Support as well. All 12 Plugins supported by Pennylane!!

arshpreetsingh commented:
> Hi @arshpreetsingh. Thanks again for your contributions. I have some notes:
> 
> I am unable to test any of the images currently as the base build is failing. I think the documentation needs to be updated to match the recent changes, following this fix so that we can build and test each image. I think it would be great if all the images can build and run the tests before the next review.
> 
> Also, the dockerfiles themselves differ a lot in indentation level. I think this should be standardised across the PR and following the examples as given in the Docker [best-practices guide](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/).

Resolved, My IDE was causing tab-space issue. Also reduced RUN calls as per the above link you had shared.

mlxd reviewed (changes requested):
Hi @arshpreetsingh thanks again for the work on this. I do have change suggestions, based on the new plugin work you have added, as well as some minor wording changes for the documentation.

thisac commented on README.md:
```diff
@@ -101,6 +101,53 @@ scripts.
+**Note:** Docker builds using "make" will work on Linux and MAC only, For MS Windows one can use [WSL](https://docs.microsoft.com/en-us/windows/wsl/install-win10)  
+
+### Build Basic Pennylane using Docker
+- To build a basic PennyLane image without any additional interfaces (Torch, TensorFlow, or Jax) and **plugins** (qiskit, amazon-bracket, cirq, rigetti) run following:
```
The `and` should probably be changed to an `or` here: `...without any additional interfaces or plugins...`. Otherwise I think this statement would be slightly confusing/misleading.

thisac commented on docker/interfaces/gpu-cuda/build_interface_gpu.dockerfile:
```diff
@@ -0,0 +1,24 @@
+# Run Unit-Tests again
+WORKDIR /opt/pennylane
+RUN make test
+# Image completed, Exit Now.
```
```suggestion
# Image build completed.
```

thisac commented on docker/interfaces/gpu-cuda/cuda-base.dockerfile:
```diff
@@ -0,0 +1,43 @@
+ENV PATH="/opt/venv/bin:$PATH"
+RUN apt-get update && apt-get install -y apt-utils \
+    --no-install-recommends python3 python3-pip python3-venv
+# Image completed, Exit Now.
```
```suggestion
# Image build completed.
```

thisac commented on docker/interfaces/gpu-cuda/build_interface_gpu.dockerfile:
```diff
@@ -0,0 +1,24 @@
+From pennylane/cuda/base:latest
```
```suggestion
FROM pennylane/cuda/base:latest
```

thisac commented on docker/pennylane.dockerfile:
```diff
@@ -0,0 +1,43 @@
+ENV PATH="/opt/venv/bin:$PATH"
+RUN apt-get update && apt-get install -y apt-utils \
+    --no-install-recommends python3 python3-pip python3-venv
+# Image completed, Exit Now.
```
```suggestion
# Image build completed.
```

thisac commented on docker/plugins/build_plugin.dockerfile:
```diff
@@ -0,0 +1,68 @@
+# Run Unit-Tests again
+WORKDIR /opt/pennylane
+RUN make test
+# Image completed, Exit Now.
```
```suggestion
# Image build completed.
```

thisac commented on docker/plugins/build_plugin.dockerfile:
```diff
@@ -0,0 +1,68 @@
+WORKDIR /opt/pennylane
+RUN make test
+# Image completed, Exit Now.
+CMD echo "Successfully built Docker image!"
```
Just to be consistent.
```suggestion
CMD echo "Successfully built Docker image"
```

thisac reviewed (commented):
Thanks for working on this @arshpreetsingh. Just adding a few minor suggested changes.

arshpreetsingh commented on docker/plugins/build_plugin.dockerfile:
```diff
@@ -0,0 +1,68 @@
+# Build Ionq Plugin
+RUN if [ "$PLUGIN_NAME" = "ionq" ] ; then \
+    apt-get update && apt-get -y install --no-install-recommends make \
+    && rm -rf /var/lib/apt/lists/* && pip3 install pennylane-ionq;fi
```
Regarding that, What if we try to do something more automated using Python?
https://docker-py.readthedocs.io/en/stable/images.html

arshpreetsingh replied:
Or more robust with Golang to handle Image-building. https://pkg.go.dev/github.com/docker/docker/client

mlxd replied:
I think for the moment it would be better to refactor the above bash scripts, rather than introducing Python/Go dependencies to the toolchain.

arshpreetsingh commented:
> Just an update: the tests for the CPU images are passing except for one. I'm currently relaying the message to the team to see about a resolution.
> 
> In the mean-time, I'd recommend merging the recent release master into here, running the images locally, and posting any logging output you get for failures here, as this may help debug the issue. Thanks!

arshpreetsingh commented:
> Hi @arshpreetsingh thanks again for the work on this. I do have change suggestions, based on the new plugin work you have added, as well as some minor wording changes for the documentation.

Hi @mlxd  thanks for Review! 
I have added required scripts for interfaces and Plugins, Also resolved other given comments. 
Looking forward for it.

mlxd commented on README.md:
````diff
@@ -101,6 +101,52 @@ scripts.
+```
+- #### To check all available interfaces run the following:
+
+```console
+make -f docker/Makefile available-interfaces
+```
+- #### To check all available plugins run the following:
+
+```console
+make -f docker/Makefile available-plugins
+```
````
I like this idea -- good job!

mlxd reviewed (commented):
Great job --- nothing more to add on my side. I'll approve, but suggest we get a second approval from the core team before merging.

thisac reviewed (changes requested):
Hi @arshpreetsingh. Thanks a lot for working on this, and on addressing all of our comments. I just had another look at it, and just noticed a few things that might be good to update. These are mainly very minor documentation change requests, to conform to the rest of our documentation. Otherwise, this looks great! 💯

thisac commented on README.md:
```diff
@@ -101,6 +101,53 @@ scripts.
+**Docker** support exists for building using **CPU** and **GPU** (Nvidia CUDA 11.1+) images.
+
+
+**Note:** Docker builds using "make" will work on Linux and MacOS only. For MS Windows you can use [WSL](https://docs.microsoft.com/en-us/windows/wsl/install-win10).
```
I've had some issues getting this to run on my M1 Mac, so I'd suggest adding a line about not supporting it for now (we could have a closer look at the issue in the future).

```suggestion
**Note:** Docker builds using "make" will work on Linux and MacOS only. For MS Windows you can use [WSL](https://docs.microsoft.com/en-us/windows/wsl/install-win10). They are currently not supported on the Apple M1 chip (ARM64).

```

arshpreetsingh replied:
@thisac  Interesting, I thought It will work on M1 as well.
Did you tried this?
https://docs.docker.com/docker-for-mac/apple-silicon/

thisac replied:
I can get it to run fine, but I'm having issues with the PennyLane installation step. There seems to be some required installations that, surprisingly, aren't there (`pkg-config` and `HDF5`, for example), and I couldn't get it to work simply by adding a step to install them. 

Unless you wish look into this a bit more, I'd suggest we leave this for now. Since everything seems to work for @mlxd, I'm happy to approve this with a note on M1 Mac support.

thisac reviewed (approved):
Thanks for all the work on this! Looks good to me now, so I'm happy to approve. 💯

thisac commented:
Thanks a lot @arshpreetsingh for this contribution. 🎉  

I've just merged it in, but I realized that the changelog hadn't been updated with these addition. It would be great if you could add a small PR just updating [the changelog](https://github.com/PennyLaneAI/pennylane/blob/master/.github/CHANGELOG.md) with what was added in this PR, along with your name in contributors list at the bottom of the development release.

arshpreetsingh commented:
> Thanks a lot @arshpreetsingh for this contribution. 
> 
> I've just merged it in, but I realized that the changelog hadn't been updated with these addition. It would be great if you could add a small PR just updating [the changelog](https://github.com/PennyLaneAI/pennylane/blob/master/.github/CHANGELOG.md) with what was added in this PR, along with your name in contributors list at the bottom of the development release.

@thisac  Please find PR here: https://github.com/PennyLaneAI/pennylane/pull/1478
